engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/PassManager.scala

```diff
@@ -165,6 +166,41 @@ class PassManager(
   ): IRType = {
     val pendingMiniPasses: ListBuffer[MiniPassFactory] = ListBuffer()
 
+    def checkDuplicates(msg: String, ir: IRType): IRType = {
+      def toString(ir: IR): String = {
+        ir.getClass().getName() + "@" + Integer.toHexString(
+          System.identityHashCode(ir)
+        ) + ":" + ir.showCode()
+      }
+
+      val all   = ir.preorder()
+      val idMap = new java.util.IdentityHashMap[Any, Boolean]()
+      val duplFound = all.find(e => {
+        if (reported.containsKey(System.identityHashCode(e))) {
+          false
+        } else {
+          if (idMap.containsKey(e)) {
```

Comment:
Scans for current `ir` in `preorder` and finds first `==` `IR` element that appears twice and hasn't yet been reported.

If such a `duplFound`, then it scans the `IR` again to find all elements that have the `duplFound.orNull` among their `children()` and prints that up.

The results of such a scan are [visible in the log](https://github.com/enso-org/enso/actions/runs/12350627142/job/34463904431?pr=11873#step:7:1473) for example like:
```
Duplicate found after org.enso.compiler.pass.desugar.FunctionBinding$: org.enso.compiler.core.ir.Name$Literal@6edf29c1:Named_Pattern
   referenced by org.enso.compiler.core.ir.module.scope.definition.Method$Conversion@3791f50e:Regex.from = (that : Named_Pattern) -> ((compile) Regex ((regex_pattern) that))
   referenced by org.enso.compiler.core.ir.DefinitionArgument$Specified@7c8c70d6:(that : Named_Pattern)
```
In some cases such a sharing seems innocent, but in others it may be dangerous. Such a dangerous sharing was probably triggered by #11365 - investigation pending.